### PR TITLE
[flutter_tools] Make ApplicationPackageFactory inject dependencies for Android Builds

### LINKED ISSUE #67818: [tool_crash] Invalid argument(s): Cannot find executable for /Users/poisson/Library/Android/sdk/build-tools/30.0.2/aapt.
## Command
```
flutter --no-color run --machine --track-widget-creation --device-id=emulator-5554 --start-paused --dart-define=flutter.inspector.structuredErrors=true lib/main.dart
```

## Steps to Reproduce
1. ...
2. ...
3. ...

## Logs
ArgumentError: Invalid argument(s): Cannot find executable for /Users/poisson/Library/Android/sdk/build-tools/30.0.2/aapt.
```
#0      _getExecutable (package:process/src/interface/local_process_manager.dart:127:5)
#1      LocalProcessManager.runSync (package:process/src/interface/local_process_manager.dart:94:30)
#2      _DefaultProcessUtils.runSync (package:flutter_tools/src/base/process.dart:417:51)
#3      new AndroidApk.fromApk (package:flutter_tools/src/application_package.dart:120:33)
#4      AndroidApk.fromAndroidProject (package:flutter_tools/src/application_package.dart:173:27)
<asynchronous suspension>
#5      ApplicationPackageFactory.getPackageForPlatform (package:flutter_tools/src/application_package.dart:47:32)
#6      FlutterDevice.runHot (package:flutter_tools/src/resident_runner.dart:409:56)
<asynchronous suspension>
#7      HotRunner.run (package:flutter_tools/src/run_hot.dart:361:31)
#8      AppDomain.startApp.<anonymous closure> (package:flutter_tools/src/commands/daemon.dart:518:23)
#9      AppDomain.launch.<anonymous closure> (package:flutter_tools/src/commands/daemon.dart:580:26)
#10     AppDomain.launch.<anonymous closure> (package:flutter_tools/src/commands/daemon.dart:578:38)
#11     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:29)
#12     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:7)
#13     _rootRun (dart:async/zone.dart:1184:13)
#14     _CustomZone.run (dart:async/zone.dart:1077:19)
#15     _runZoned (dart:async/zone.dart:1619:10)
#16     runZoned (dart:async/zone.dart:1539:10)
#17     AppContext.run (package:flutter_tools/src/base/context.dart:149:18)
```
```
[✓] Flutter (Channel stable, v1.17.4, on Mac OS X 10.15.7 19H2, locale zh-Hans-CN)
    • Flutter version 1.17.4 at /Users/poisson/flutter
    • Framework revision 1ad9baa8b9 (4 months ago), 2020-06-17 14:41:16 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/poisson/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    ✗ Android SDK file not found: /Users/poisson/Library/Android/sdk/build-tools/30.0.2/aapt.
    • Try re-installing or updating your Android SDK,
      visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0, Build version 12A7209
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 48.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[!] IntelliJ IDEA Ultimate Edition (version 2020.1.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] Connected device (4 available)
    • sdk gphone x86 arm • emulator-5554                        • android-x86 • Android 11 (API 30) (emulator)
    • iPhone             • 00008020-00142599227A002E            • ios         • iOS 14.0.1
    • iPhone 8           • 770CB6AF-96D4-4891-BB41-B72475F33310 • ios         • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)
    • iPhone 11          • 51E96ED3-3947-4CE1-B203-FA50BA972EFC • ios         • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)

! Doctor found issues in 2 categories.

```

## Flutter Application Metadata
**Type**: app
**Version**: 1.0.0+1
**Material**: true
**Android X**: false
**Module**: false
**Plugin**: false
**Android package**: null
**iOS bundle identifier**: null
**Creation channel**: stable
**Creation framework version**: 1ad9baa8b99a2897c20f9e6e54d3b9b359ade314
### Plugins
fluttertoast-7.0.4
path_provider-1.6.14
path_provider_linux-0.0.1+2
path_provider_macos-0.0.4+3
shared_preferences-0.5.10
shared_preferences_linux-0.0.2+2
shared_preferences_macos-0.0.1+10
shared_preferences_web-0.1.2+7
sqflite-1.3.1+1
webview_flutter-0.3.22+1

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Remove globals and fix case where we did not check if appt was available, leading to a tool crash.

Fixes https://github.com/flutter/flutter/issues/67818

### CODE REVIEW

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 1]
```diff
@@ -577,7 +577,8 @@ class AndroidDevice extends Device {
       );
       // Package has been built, so we can get the updated application ID and
       // activity name from the .apk.
-      package = await AndroidApk.fromAndroidProject(project.android);
+      package = await ApplicationPackageFactory.instance
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/android_device.dart:580]: eventually this should be passed as an argument or injected, but I did not want to refactor all of the device types at once

File: packages/flutter_tools/lib/src/application_package.dart [hunk 2]
```diff
@@ -107,10 +147,18 @@ class AndroidApk extends ApplicationPackage {
...
+  /// Returns `null` if the APK was invalid or any required tooling was missing.
+  factory AndroidApk.fromApk(File apk, {
+    @required AndroidSdk androidSdk,
+    @required ProcessManager processManager,
+    @required UserMessages userMessages,
+    @required ProcessUtils processUtils,
+    @required Logger logger,
+  }) {
+    final String aaptPath = androidSdk?.latestVersion?.aaptPath;
+    if (aaptPath == null || !processManager.canRun(aaptPath)) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/application_package.dart:160]: This was the change that prompted the investigation...

File: packages/flutter_tools/lib/src/application_package.dart [hunk 3]
```diff
@@ -412,52 +474,6 @@ class PrebuiltIOSApp extends IOSApp {
   String get _bundlePath => bundleDir.path;
 }
 
-class ApplicationPackageStore {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/application_package.dart:415]: This does _almost_ the same thing as application package factory, it is just older and used for drive. It does not seem good to have two _almost_ the same classes

File: packages/flutter_tools/test/src/mocks.dart [hunk 4]
```diff
@@ -35,31 +33,6 @@ final Generator kNoColorTerminalPlatform = () {
   )..stdoutSupportsAnsi = false;
 };
 
-class MockApplicationPackageStore extends ApplicationPackageStore {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/src/mocks.dart:38]: Deleting these had no effect, except on drive which just need one fake

File: packages/flutter_tools/lib/src/android/android_device.dart (hunk 1 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/android/android_device.dart:578]: SGTM

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**